components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt

```diff
@@ -0,0 +1,277 @@
+         */
+        internal fun fromJsonString(json: String): TimingDistributionData? {
+            val jsonObject: JSONObject
+            try {
+                jsonObject = JSONObject(json)
+            } catch (e: org.json.JSONException) {
+                return null
+            }
+
+            // Category can be empty so it may be possible to be a null value so try and allow this
+            // by using `orEmpty()` to fill in the value.  Other values should be present or else
+            // something is wrong and we should return null.
+            val category = jsonObject.tryGetString("category").orEmpty()
+            val name = jsonObject.tryGetString("name") ?: return null
+            val bucketCount = jsonObject.tryGetInt("bucketCount") ?: return null
+            // If 'range' isn't present, JSONException is thrown
+            val range = try {
+                jsonObject.getJSONArray("range").toList<Long>()
+            } catch (e: org.json.JSONException) {
+                return null
+            }
+            val histogramType = HistogramType.values()[jsonObject.tryGetInt("histogramType")
+                ?: return null]
+            // Attempt to parse the values map, if it fails then something is wrong and we need to
+            // return null.
+            val values = try {
+                val mapData = jsonObject.getJSONObject("values")
+                val valueMap: MutableMap<String, Long> = mutableMapOf()
+                mapData.keys().forEach { key ->
+                    valueMap[key] = mapData.tryGetLong(key) ?: 0L
+                }
+                valueMap
+            } catch (e: org.json.JSONException) {
+                // This should only occur if there isn't a key/value pair stored for "values"
+                return null
+            }
+            val sum = jsonObject.tryGetLong("sum") ?: return null
+            val timeUnit = TimeUnit.values()[jsonObject.tryGetInt("timeUnit")
+                ?: return null]
+
+            return TimingDistributionData(
+                category = category,
+                name = name,
+                bucketCount = bucketCount,
+                range = range,
+                histogramType = histogramType,
+                values = values,
+                sum = sum,
+                timeUnit = timeUnit
+            )
+        }
+    }
+
+    // This is a helper property to build the correct identifier for the metric and allow for
+    // blank categories
+    internal val identifier: String = if (category.isEmpty()) { name } else { "$category.$name" }
+
+    // This is a list of limits for the buckets.  Instantiated lazily to ensure that the range and
+    // bucket counts are set first.
+    internal val buckets: List<Long> by lazy { getBuckets() }
+
+    /**
+     * Accumulates a sample to the correct bucket
+     *
+     * @param sample Long value representing the sample that is being accumulated
+     */
+    internal fun accumulate(sample: Long) {
```

Comment:
Can we add test coverage for this specific function, to make sure that the search below is correct? (this is not the same of testing the caller)